Error while checking build/O2/fullCI for 0ba2f884b61aab6edd5bd9a25614ee0a4bb9131f at 2023-01-19 01:36:
```
## sw/BUILD/o2checkcode-latest/log
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/Common/test/testGPUsortCUDA.cu:22:10: error: 'boost/test/unit_test.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstruction.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstructionCPU.cxx:45:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUTracking/display/GPUDisplay.cxx:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUTracking/Base/cuda/GPUReconstructionCUDAGenRTC.cu:16:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/benchmark.hip.cxx:199:35: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:8: error: unknown type name '__host__' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:27: error: expected ';' after top level declarator [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:33: error: overloaded 'operator+=' must have at least one parameter of class or enumeration type [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:44: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:53: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:85:1: error: unknown type name '__global__' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:19: error: use of undeclared identifier 'blockIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:32: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:45: error: use of undeclared identifier 'threadIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:78: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10566-slc8_x86-64/0/Detectors/TOF/calibration/src/TOFChannelCalibrator.cxx:23:10: error: 'omp.h' file not found [clang-diagnostic-error]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10566/0ba2f884b61aab6edd5bd9a25614ee0a4bb9131f/build_O2_fullCI/pretty.html).

--- a/Detectors/MUON/MCH/Calibration/include/MCHCalibration/PedestalData.h
+++ b/Detectors/MUON/MCH/Calibration/include/MCHCalibration/PedestalData.h
@@ -133,25 +133,36 @@ class PedestalDataIterator
     return &mMapIt->second[mRow][mCol];
   }
 
-  PedestalDataIterator& operator++()
+  bool advance()
   {
     if (mMapIt != mData->mPedestals.end()) {
       if (mCol < PedestalData::MAXCHANNEL - 1) {
         mCol++;
-        return *this;
+        return true;
       }
       if (mRow < PedestalData::MAXDS - 1) {
         mCol = 0;
         mRow++;
-        return *this;
+        return true;
       }
       ++mMapIt;
       mCol = 0;
       mRow = 0;
       if (mMapIt != mData->mPedestals.end()) {
-        return *this;
+        return true;
       }
       mData = nullptr;
+    }
+    // undefined behavior here (should not increment an end iterator)
+    return false;
+  }
+
+  PedestalDataIterator& operator++()
+  {
+    while (advance()) {
+      if ((*this)->isValid()) {
+        break;
+      }
     }
     // undefined behavior here (should not increment an end iterator)
     return *this;

--- a/Detectors/MUON/MCH/Calibration/src/BadChannelCalibrator.cxx
+++ b/Detectors/MUON/MCH/Calibration/src/BadChannelCalibrator.cxx
@@ -71,7 +71,7 @@ bool BadChannelCalibrator::hasEnoughData(const Slot& slot) const
   const int requiredChannels = static_cast<int>(BadChannelCalibratorParam::Instance().minRequiredCalibratedFraction * nofChannels);
 
   auto nofCalibrated = std::count_if(pedData->cbegin(), pedData->cend(),
-                                     [&](const PedestalChannel& c) { return (c.isValid() && (c.mEntries > minNofEntries)); });
+                                     [&](const PedestalChannel& c) { return c.mEntries > minNofEntries; });
 
   bool hasEnough = nofCalibrated > requiredChannels;
 
@@ -100,7 +100,7 @@ void BadChannelCalibrator::finalizeSlot(Slot& slot)
   }
 
   for (const auto& ped : *pedestalData) {
-    if (!ped.isValid() || ped.mEntries == 0) {
+    if (ped.mEntries == 0) {
       continue;
     }
     mPedestalsVector.emplace_back(ped);

--- a/Detectors/MUON/MCH/Calibration/test/example-mch-pedestal-calibration.sh
+++ b/Detectors/MUON/MCH/Calibration/test/example-mch-pedestal-calibration.sh
@@ -1,7 +1,10 @@
+date
+
 o2-raw-tf-reader-workflow --input-data \
-    $HOME/alice/data/2022/MAY/514781/raw/1100 --onlyDet MCH --max-tf 10 | \
+    $HOME/alice/data/2022/LHC22t_MCH/529437/calib/1140 --onlyDet MCH --max-tf 1000 | \
 o2-mch-pedestal-decoding-workflow | \
-o2-calibration-mch-badchannel-calib-workflow | \
+o2-calibration-mch-badchannel-calib-workflow \
+  --configKeyValues="MCHBadChannelCalibratorParam.minRequiredNofEntriesPerChannel=10;MCHBadChannelCalibratorParam.minRequiredCalibratedFraction=0.5;MCHBadChannelCalibratorParam.onlyAtEndOfStream=true" | \
 o2-calibration-ccdb-populator-workflow \
     --ccdb-path http://localhost:6464 \
     --sspec-min 0 --sspec-max 0 | \
@@ -9,3 +12,4 @@ o2-calibration-ccdb-populator-workflow \
     --ccdb-path localhost:8484 --sspec-max 1 --sspec-min 1 --name-extention dcs | \
 o2-dpl-run --run -b
 
+date

--- a/Detectors/MUON/MCH/Calibration/test/testPedestalData.cxx
+++ b/Detectors/MUON/MCH/Calibration/test/testPedestalData.cxx
@@ -13,10 +13,13 @@
 #define BOOST_TEST_DYN_LINK
 #include <boost/test/unit_test.hpp>
 
-#include "MCHCalibration/PedestalDigit.h"
 #include "MCHCalibration/PedestalData.h"
-#include <vector>
+#include "MCHCalibration/PedestalDigit.h"
+#include "MCHConstants/DetectionElements.h"
+#include "MCHMappingInterface/Segmentation.h"
+#include "MCHRawElecMap/Mapper.h"
 #include <random>
+#include <vector>
 
 std::vector<uint16_t> samples(int n)
 {
@@ -35,14 +38,53 @@ using o2::mch::calibration::PedestalDigit;
 struct PedestalDigits {
   PedestalDigits()
   {
+    // non existing DS : that channel should not appear when iterating on pedestal data
     digits.emplace_back(721, 23, 13, 1234, 43, samples(13));
+
+    // solarId 721 groupId 5 indexId 1
+    // S721-J5-DS1 (elinkId 26) [ FEE30-LINK1 ] DE708-DS102
     digits.emplace_back(721, 26, 16, 2345, 46, samples(16));
-    digits.emplace_back(328, 28, 18, 3456, 48, samples(18));
+
+    // solarId 328 groupId 5 indexId 2
+    // S328-J5-DS2 (elinkId 27) [ FEE32-LINK6 ] DE714-DS108 channel 18
+    digits.emplace_back(328, 27, 18, 3456, 48, samples(18));
   }
   std::vector<PedestalDigit> digits;
 };
 
 BOOST_FIXTURE_TEST_SUITE(PedestalDataIterator, PedestalDigits)
+
+BOOST_AUTO_TEST_CASE(TestIteratorOnCompletePedestalData)
+{
+  // build a vector of all possible digits
+  std::vector<PedestalDigit> allDigits;
+
+  auto det2elec = o2::mch::raw::createDet2ElecMapper<o2::mch::raw::ElectronicMapperGenerated>();
+
+  for (auto deId : o2::mch::constants::deIdsForAllMCH) {
+    const auto& seg = o2::mch::mapping::segmentation(deId);
+    seg.forEachPad([&](int padID) {
+      if (seg.isValid(padID)) {
+        auto dsId = seg.padDualSampaId(padID);
+        auto ch = seg.padDualSampaChannel(padID);
+        o2::mch::raw::DsDetId det(deId, dsId);
+        auto elec = det2elec(det).value();
+        auto solarId = elec.solarId();
+        allDigits.emplace_back(solarId, elec.elinkId(), ch, 42, 42, samples(15));
+      }
+    });
+  }
+
+  PedestalData pd;
+  pd.fill(allDigits);
+
+  BOOST_REQUIRE(allDigits.size() == 1063528);
+  int n{0};
+  for (const auto& ped : pd) {
+    ++n;
+  }
+  BOOST_TEST(n == allDigits.size());
+}
 
 BOOST_AUTO_TEST_CASE(TestIteratorEquality)
 {
@@ -70,7 +112,23 @@ BOOST_AUTO_TEST_CASE(TestIteratorPreIncrementable)
   for (auto rec : pd) {
     n++;
   }
-  BOOST_TEST(n == PedestalData::MAXDS * PedestalData::MAXCHANNEL * 2);
+  BOOST_TEST(n == 2768);
+  // 2768 = 1856 pads in solar 328 + 721 pads in solar 721
+  // Note that solar 328 has 29 dual sampas
+  // solar 721 has 15 dual sampas
+  // But 2768 < (29+15)*64 (=2816) because not all pads are valid ones.
+}
+
+BOOST_AUTO_TEST_CASE(TestIteratorAllReturnedPadAreValidByConstruction)
+{
+  PedestalData pd;
+  pd.fill(digits);
+  auto n = std::count_if(pd.begin(), pd.end(), [](o2::mch::calibration::PedestalChannel& c) {
+    return c.isValid();
+  });
+  auto n1 = std::distance(pd.begin(), pd.end());
+  BOOST_TEST(n == 2768);
+  BOOST_CHECK(n == n1);
 }
 
 BOOST_AUTO_TEST_CASE(TestIteratorInCountIfAlgorithm)
@@ -80,7 +138,7 @@ BOOST_AUTO_TEST_CASE(TestIteratorInCountIfAlgorithm)
   auto n = std::count_if(pd.begin(), pd.end(), [](o2::mch::calibration::PedestalChannel& c) {
     return c.mEntries > 0;
   });
-  BOOST_TEST(n == 3);
+  BOOST_TEST(n == 2); // 2 and not 3 because one of the digit is on a pad that is not connected (hence invalid, hence not part of the iteration)
 }
 
 BOOST_AUTO_TEST_CASE(IterationOnEmptyDataShouldNotBeAnInfiniteLoop, *boost::unit_test::timeout(10))

